lib/pbench/test/unit/server/test_indexing_tarballs.py

```diff
@@ -0,0 +1,317 @@
+        assert operation in __class__.tarballs
+        return __class__.tarballs[operation]
+
+    def update(
+        self,
+        dataset: Dataset,
+        did: Optional[Operation],
+        enabled: Optional[List[Operation]],
+    ):
+        __class__.did = did
+        __class__.updated = enabled
+
+    def error(self, dataset: Dataset, message: str):
+        __class__.errors = message
+
+
+class FakeController:
+    def __init__(self, path: Path, incoming: Path, results: Path, logger: Logger):
+        self.name = path.name
+        self.path = path
+        self.incoming = incoming / self.name
+        self.results = results / self.name
+        self.logger = logger
+
+
+class FakeTarball:
+    def __init__(self, path: Path, controller: FakeController):
+        self.name = path.name
+        self.tarball_path = path
+        self.controller = controller
+        self.unpacked = f"/incoming/{path.name}"
+
+
+class FakeCacheManager:
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        self.config = config
+        self.logger = logger
+        self.datasets = {}
+
+    def find_dataset(self, resource_id: str):
+        controller = FakeController(
+            Path("/archive/ctrl"), Path("/incoming"), Path("/results"), self.logger
+        )
+        return FakeTarball(
+            Path(f"/archive/ctrl/tarball-{resource_id}.tar.xz"), controller
+        )
+
+
+@pytest.fixture()
+def mocks(monkeypatch, make_logger):
+    FakeDataset.logger = make_logger
+    with monkeypatch.context() as m:
+        m.setattr("pbench.server.indexing_tarballs.Sync", FakeSync)
+        m.setattr("pbench.server.indexing_tarballs.PbenchTarBall", FakePbenchTarBall)
+        m.setattr("pbench.server.indexing_tarballs.Report", FakeReport)
+        m.setattr("pbench.server.indexing_tarballs.Dataset", FakeDataset)
+        m.setattr("pbench.server.indexing_tarballs.Metadata", FakeMetadata)
+        m.setattr("pbench.server.indexing_tarballs.CacheManager", FakeCacheManager)
+        yield m
+    FakeSync.reset()
+    FakePbenchTarBall.reset()
+
+
+@pytest.fixture()
+def index(server_config, make_logger):
+    return Index(
+        "test",
+        Namespace(index_tool_data=False, re_index=False),
+        FakeIdxContext(server_config, make_logger),
+    )
+
+
+class TestIndexingTarballs:
+    def test_construct(self, mocks, index, server_config):
+        assert not index.options.index_tool_data
+        assert index.sync.component == "index"
+
+    def test_collect_tb_empty(self, mocks, index):
+        FakeSync.tarballs[Operation.INDEX] = []
+        tb_list = index.collect_tb()
+        assert FakeSync.called == ["next-INDEX"]
+        assert tb_list == (0, [])
+
+    def test_collect_tb(self, mocks, index):
+        def stat(file: str) -> stat_result:
+            assert "ds1" in file or "ds2" in file
+            size = 5 if "ds1" in file else 2
+            return stat_result([0o777, 123, 300, 1, 100, 100, size, 0, 0, 0])
```

Comment:
> I'm hotwiring the mock to be able to inject an exception, so lambda is off the table.

I'm not sure what you're saying.  The code cited above will raise an `AssertionError` under certain circumstances.  My suggestion will raise a `KeyError` in that case (assuming that we can finesse the filename/path problem).  Either way, the exception gets injected, and my way can do it from a lambda expression.  🙂 

Nevertheless, you're quite correct that replacing `size` with `sizes[Dataset.stem(file)]` prompts Black to wrap the lines, which is sad.

The lambda expression creates a bit of a spread (although not a huge one), e.g.:
```
	mocks.setattr(
            "os.stat",
            lambda file: stat_result(
                [0o777, 123, 300, 1, 100, 100, sizes[Dataset.stem(file)], 0, 0, 0]
            ),
        )
```
I'm not sure how readable that is, but the immediacy does have some value.  Otherwise, I guess the best result is to use a local variable:
```
        def stat(file: str) -> stat_result:
            size = sizes[Dataset.stem(file)]
            return stat_result([0o777, 123, 300, 1, 100, 100, size, 0, 0, 0])
        ...
	mocks.setattr("os.stat", stat)
```
which is one line shorter than the lambda, but at the cost of placing the definition for the mock apart from what it's replacing.